### ISSUE TITLE
Support TLS key logging

### ISSUE THREAD

**Issue by seidnerj** [ProxymanApp/Proxyman#1671]
## Description

Support TLS key logging, this can be implemented by Proxyman looking for a "SSLKEYLOGFILE" environment variable as is done in other applications or by some other means (UI-wise).

## Why this feature/change is important?

I am trying to debug issues that seem to origin in the communicate between Proxyman and the target server, I am using Wireshark to look at the relevant traffic but unfortunately (in this case) it is encrypted (TLS), if the keys could be logged as in other applications, I could then load them to Wireshark and take a look at the unencrypted traffic. Without this, I have no insight into what is happening between Proxyman and the target server.

**Comment by swznd** [comment]
agreed, I have similar issue, I need to compare request between app to proxyman and proxyman to target server

**Comment by dabing1022** [comment]
June 1, 2024. Excuse me, is it supported now?

**Comment by novitae** [comment]
Any news ? @NghiaTranUIT

**Comment by NghiaTranUIT** [comment]
@novitae To be honest, I don't know how to implement it. Can you show me the current solution from another app?

**Comment by novitae** [comment]
> @novitae To be honest, I don't know how to implement it. Can you show me the current solution from another app?

I just took a look at [how mitmproxy does it](https://github.com/mitmproxy/mitmproxy/blob/70e59c2d4157f017f09163bc6b33bcd2a8f83594/mitmproxy/net/tls.py#L119C1-L164C1), but their backend is python, so it will be different for you.

What crypto backend are you using to relay the requests ? OpenSSL ? If it is, I'm pretty sure it's easy and there are already methods to make it easier

**Comment by NghiaTranUIT** [comment]
@novitae is it what you're looking for?

I add your suggestion to the SSLContext between clients -> Proxyman (not Proxyman -> Server). From mitmproxy code, it applies to 2 path, client -> mitmproxy, and mitmproxy to server.


![Screenshot 2025-01-08 at 14 51 49](https://github.com/user-attachments/assets/0c5a68cc-6625-42bc-b107-cee865c71aec)

**Comment by NghiaTranUIT** [comment]
It's what you're looking for?

![Screenshot 2025-01-08 at 7 43 42 PM](https://github.com/user-attachments/assets/46bf6fce-95fa-496c-860c-a719054f4c19)

- mitmproxy does TLSKeyLogging for both way: client -> mitmproxy, and mitmproxy -> servers. Do you need both?

**Comment by NghiaTranUIT** [comment]
thanks. I will add it. What do you do with this file? These secrets don't link with certain domains, so It's hard to know which domain SSL is.

**Comment by novitae** [comment]
> thanks. I will add it. What do you do with this file? These secrets don't link with certain domains, so It's hard to know which domain SSL is.

Yes I must admit the way it works I a bit confusing, and I couldn't explain to you how it works to find which key is for which connection, but read this you might understand better:

https://wiki.wireshark.org/TLS#Using%20the%20(Pre)-Master-Secret

**Comment by NghiaTranUIT** [comment]
@novitae @dabing1022 @swznd @seidnerj Good news: TLS Key Logging is officially supported on this Beta build: https://download.proxyman.io/beta/Proxyman_5.12.2_Support_TLS_Key_Logging.dmg

You can access it in the Tool menu -> TLS Key Logging -> Select your File or Folder.

<img width="617" alt="Screenshot 2025-01-11 at 11 08 12 AM" src="https://github.com/user-attachments/assets/7cdf0d3b-297b-4c55-9b01-cc0e28ca1703" />

- I prefer using UI because macOS app can't read your system env `SSLKEYLOGFILE`

**Comment by novitae** [comment]
Thank you ! I just tried it and decrypted with Wireshark, it works perfectly

**Comment by seidnerj** [comment]
This is awesome!!! Thank you so much @NghiaTranUIT! 🙏
